# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abrimos los datos y los examinamos

Necesitarás `pandas`, así que impórtalo.

In [1]:
import pandas as pd# Importar pandas


Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
df= pd.read_csv("/datasets/music_project_en.csv")# Leer el archivo y almacenarlo en

Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
# Obtener la información general sobre nuestros datos
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. se podria dejar nombres de las columnas un poco mas explicado o mucho mas claro, ya que dia puede ser cual quie dia mas bien `'dayofweek'` y en el caso de `'time'` se puede dejar como `'totalplay'` puede llegar a ser un poco confuso.




### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`tenemos información de tipo object dentro de cada columna. De igual manera podemos ver los valores únicos que se encuentran dentro cada columna así como la cantidad de los valores únicos que hay, así como también la cantidad de valores repetido que se encuentran en cada columna. Detectar si hay presente valores nulos y mirar cuantos hay, en que tiempo se presenta valores nulos. 

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`de pronto seria bueno la fecha completa mes y año, porque tambien puede varias las canciones escuchadas en relación a la temporada del año o tambien podriamos tener información en relación a la temporada del año, si es invierno, verano o otoño

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?` se puede evidencia que hay datos ausentes en las columnas `'Track'`, `'artist'` y `'genre'` y se presenta que hay 3826 datos duplicdos

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [6]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [7]:
# Bucle en los encabezados poniendo todo en minúsculas
new_name=[old_name.lower() for old_name in df.columns]
print(new_name)

['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day']


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [8]:
# Bucle en los encabezados eliminando los espacios
new_name=[old_name.lower() for old_name in df.columns]
name_lowered=[name.strip() for name in new_name]
print(name_lowered)

['userid', 'track', 'artist', 'genre', 'city', 'time', 'day']


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [9]:
# Cambiar el nombre de la columna "userid"
df.columns= name_lowered
df.rename(columns={'userid':'user_id'},inplace=True)

Comprueba el resultado. Muestra los encabezados una vez más:

In [10]:
# Comprobar el resultado: la lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [11]:
# Calcular el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [12]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
Columns_to_replace=['track', 'artist', 'genre']
for col in Columns_to_replace:
    df[col].fillna('unknown', inplace=True)

Se comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [13]:
# Contar valores ausentes
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [14]:
# Contar duplicados explícitos
print(df.duplicated().sum())

3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [15]:
# Eliminar duplicados explícitos
print(df.drop_duplicates(inplace=True))

None


Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [16]:
# Comprobar de nuevo si hay duplicados
print(df.duplicated())
print(df.duplicated().sum())

0        False
1        False
2        False
3        False
4        False
         ...  
65074    False
65075    False
65076    False
65077    False
65078    False
Length: 61253, dtype: bool
0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [17]:
# Inspeccionar los nombres de géneros únicos
print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hip' 'jazz' 'postrock' 'latin'
 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental' 'rusrock'
 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'hiphop' 'drum' 'extrememetal' 'türkçe'
 'experimental' 'easy' 'metalcore' 'modern' 'argentinetango' 'old' 'swing'
 'breaks' 'eurofolk' 

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [18]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(df, column, wron_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df[column]=df[column].replace(wrong_genre,correct_genre)
    return df

In [19]:
# Eliminar duplicados implícitos
wrong_genres=['hip','hop','hip-hop']
correct_genre='hiphop'
types_genre= replace_wrong_genres(df,'genre',wrong_genres,correct_genre)

In [22]:
# Comprobación de duplicados implícitos
print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'drum' 'extrememetal' 'türkçe' 'experimental' 'easy'
 'metalcore' 'modern' 'argentinetango' 'old' 'swing' 'breaks' 'eurofolk'
 'stone

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

primero que todo para que podamos realizar el cumplimiento de la hipotesis o poder saber si es verdad lo que se plantea, primero debemos abarcar el tratamiento de lo datos en los cuales debemos organizarlos y limpiarlos para la posterior toma de decisiones, ya que si lo hacemos de manera incorrecta podremos tomar una mala decisión. Tambien abordamos los duplicados identificandolos en donde se encuentran y poseterior a ello, pudimos eliminarlos, reduciendo la cantidad de datos a los que son reales y no inflando valores

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


In [23]:
# Contar las canciones reproducidas en cada ciudad
print(df.groupby(by='city')['user_id', 'track'].count())

             user_id  track
city                       
Shelbyville    18512  18512
Springfield    42741  42741


se reprodujeron en **Shelbyville** un total de 18512 canciones y en el caso de **Springfield** un total de 42741 canciones

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [24]:
# Calcular las canciones reproducidas en cada uno de los tres días
print(df.groupby(by=['city','day'])['user_id', 'track'].count())

                       user_id  track
city        day                      
Shelbyville Friday        5895   5895
            Monday        5614   5614
            Wednesday     7003   7003
Springfield Friday       15945  15945
            Monday       15740  15740
            Wednesday    11056  11056


De acuerdo a los resultados obtenidos, se puede observar que el dia que menos se escucharon canciones el lunes para **Shelbyville** con 5614 y para **Springfield** fue el dia miercoles con 11056 y los dias que mas se escucharon canciones fueron para **Shelbyville** fue el miercoles con 7003 y para **Springfield** fue el dia viernes con 15945

In [25]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(df, day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    df_filter=df[df['day']== day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    df_filter=df_filter[df_filter['city']== city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    # Devolve el número de valores de la columna 'user_id'
    return df_filter['user_id'].count()

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [26]:
# El número de canciones reproducidas en Springfield el lunes
day='Monday'
city='Springfield'
songs = number_tracks(df, day, city)
print(songs)

15740


In [27]:
# El número de canciones reproducidas en Shelbyville el lunes
day='Monday'
city='Shelbyville'
songs = number_tracks(df, day, city)
print(songs)

5614


In [28]:
# El número de canciones reproducidas en Springfield el miércoles
day='Wednesday'
city='Springfield'
songs = number_tracks(df, day, city)
print(songs)

11056


In [29]:
# El número de canciones reproducidas en Shelbyville el miércoles
day='Wednesday'
city='Shelbyville'
songs = number_tracks(df, day, city)
print(songs)

7003


In [30]:
# El número de canciones reproducidas en Springfield el viernes
day='Friday'
city='Springfield'
songs = number_tracks(df, day, city)
print(songs)

15945


In [31]:
# El número de canciones reproducidas en Shelbyville el viernes
day='Friday'
city='Shelbyville'
songs = number_tracks(df, day, city)
print(songs)

5895


**Conclusiones**

la hipótesis es correcta, pero también debemos tener en cuenta otras variables, como la población de cada ciudad, así como tambien la edad de los usuarios que escucharon música. Aunque la hipótesis es verdadera el resultado está sesgado a mas variables.

# Conclusiones <a id='end'></a>

1. se escucha mas musica en springfiel que en Shelbyville
2. el dia que menos se escucha musica en Shelbyville es el lunes y en Springfield es el miercoles
3. el dia que mas se escucha musica en Shelbyville es el miercoles y en Springfield es el viernes